### ISSUE TITLE
SPIKE: Can we ship / expose DataHub logs to Cloud Platform?

### ISSUE THREAD

**Issue by jemnery** [ministryofjustice/find-moj-data#229]
### Background
There is existing logging from datahub into kibana via standard out, however it isn't clear exactly what is logged.

### Key Questions to answer
- What scenarios do we want to log
    - Metadata Changes, and what user made them
    - User account changes
    - Changes to ingestion pipelines
- Is it possible to expose the required logs to cloud platform
    - Yes, the `stdout` logs from DataHub's pods are piped to CP's monitoring elasticsearch cluster and are [searchable via kibana](https://kibana.cloud-platform.service.justice.gov.uk/)
- What changes need to be made if any
    - None
- What limitations exist
    - Users are not directly tied to Metadata changes
    - Logs are stored for 30 days
    - Larger log statements are cut off in kibana
    - The Entity Events API DataHub have made is only available in managed DataHub

### Hypothesis
Can we ship / expose DH logs to CP / Kibana to consolidate our security logs? 
Would be good for future SOC requirements.
Outcome: assess whether we can have an audit trail of actions by users in one place

### Conclusion
From my looking at the logs it appears to me that it would be easier to track user behaviour via the frontend logs rather than the backend datahub logs, which are hard to parse.

[The entity events api has event logs like we'd want, but it's only available in managed DataHub.](https://datahubproject.io/docs/managed-datahub/datahub-api/entity-events-api/)

It is still possible to get an idea of user behaviour via the datahub logs when usage is low enough that timestamps can act as a link between metadata changes and frontend interactions.

### Scenarios

- [ ] User account creation
- [x] User role changes
- [x] Changes to metadata
- [x] Adding and deleting entries
- [x] Editing tags
- [x] Editing the glossary
- [x] Editing domains
- [x] Modifying ingestions

**Comment by MatMoore** [comment]
Anything that is going to stdout is already being collected: https://user-guide.cloud-platform.service.justice.gov.uk/documentation/logging-an-app/log-collection-and-storage.html#log-collection

Example: https://kibana.cloud-platform.service.justice.gov.uk/_plugin/kibana/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-15m,to:now))&_a=(columns:!(kubernetes.pod_name,kubernetes.labels.app,kubernetes.container_image,kubernetes.namespace_name,log,stream),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'167701b0-f8c0-11ec-b95c-1d65c3682287',key:kubernetes.container_name,negate:!f,params:(query:datahub-gms),type:phrase),query:(match_phrase:(kubernetes.container_name:datahub-gms))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'167701b0-f8c0-11ec-b95c-1d65c3682287',key:kubernetes.namespace_name,negate:!f,params:(query:data-platform-datahub-catalogue-test),type:phrase),query:(match_phrase:(kubernetes.namespace_name:data-platform-datahub-catalogue-test)))),index:'167701b0-f8c0-11ec-b95c-1d65c3682287',interval:auto,query:(language:kuery,query:''),sort:!())

**Comment by YvanMOJdigital** [comment]
Threat Model workshop: https://miro.com/app/board/uXjVKRSefrc=/

**Comment by MatMoore** [comment]
Some example logs we can search for

## Role changed

```
INFO  c.l.m.k.MetadataChangeLogProcessor:102 - Invoking MCL hooks for urn: ***, aspect name: roleMembership, entity type: corpuser, change type: UPSERT

2024-05-14 15:16:00,536 [Thread-4116] INFO  c.l.m.entity.EntityServiceImpl:1514 - Producing MCL for ingested aspect roleMembership, urn ***
```

These do not log who performed the action.

## Usage events
This is triggered when I perform an action

Maybe we can consume from this queue directly and log events that way

```
INFO  c.l.m.k.DataHubUsageEventsProcessor:58 - Got DHUE event key: urn:li:corpuser:Mat.Moore, topic: DataHubUsageEvent_v1, partition: 0, offset: 12554, value size: 641, timestamp: 1715699675851
```

**Comment by MatMoore** [comment]
I created two saved searches in kibana:

- Anything on datahub dev with "Error" or "warning" in
- Anything on datahub dev with "Ingesting aspects batch to database"

The latter catches all kinds of update I've tested so far, but does distinguish between automated/manual updates and does not capture the currently logged in user.

Possibly DataHubUsageEventsProcessor will be more useful for auditing so will check what we can do with that next.

**Comment by murdo-moj** [comment]
https://datahubproject.io/docs/how/extract-container-logs/

> Log files will live inside the container under the following directories for each service:
> 
> - datahub-gms: /tmp/datahub/logs/gms
> - datahub-frontend: /tmp/datahub/logs/datahub-frontend
> 
> There are 2 types of logs that are collected:
> 
> - Info Logs: These include info, warn, error log lines. They are what print to stdout when the container runs.
> - **Debug Logs: These files have shorter retention (past 1 day)** but include more granular debug information from the DataHub code specifically. We ignore debug logs from external libraries that DataHub depends on.

**Comment by murdo-moj** [comment]
![Image](https://github.com/ministryofjustice/find-moj-data/assets/109604278/d5e99a95-1fa2-4c1a-bbb7-22fe46f7edd0)


2024-05-15 16:46:32.861 | data-platform-datahub-catalogue-test | datahub-gms | 2024-05-15 15:46:32,861 [ThreadPoolTaskExecutor-1] INFO  c.l.m.s.e.update.ESBulkProcessor:82 - Added request id: CreateDomainEvent_urn%3Ali%3Acorpuser%3AMurdo.Moyse1_1715787992784_05796, operation type: CREATE, index: datahub_usage_event
-- | -- | -- | --

**Comment by murdo-moj** [comment]
When I add a tag to a field of a dataset, logs are made that a field is being upserted. 
Because the updated aspect is editableSchemaMetadata it is requested from the UI. 
However, no user information is logged to tie that action to a user. 

```
2024-05-16 09:27:04,146 [ThreadPoolTaskExecutor-1] INFO  c.l.m.s.e.update.ESBulkProcessor:82 - Added request id: PageViewEvent_urn%3Ali%3Acorpuser%3AMurdo.Moyse1_1715851624070_05895, operation type: CREATE, index: datahub_usage_event
2024-05-16 09:27:25,077 [Thread-11788] INFO  c.l.m.entity.EntityServiceImpl:634 - Ingesting aspects batch to database, items: [ChangeItemImpl{changeType=UPSERT, urn=urn:li:dataset:(urn:li:dataPlatform:athena,opg_sirius_derived.cases_attorneys,PROD), aspectName='editableSchemaMetadata', recordTemplate={editableSchemaFieldInfo=[{globalTags={tags=[{tag=urn:li:tag:test-for-logs}]}, fieldPath=caseitem_id}]}, systemMetadata={properties={appSource=ui}}}]
```

How the logs seem to work is that user's requests to the frontend are logged against their user, but if they request a metadata change, this is not tagged to their user.

**Comment by murdo-moj** [comment]
The opensearch cluster can be called from the gms pod, eg
`curl 10.100.152.177:8080/_cat/indices?v` (get the hostname and the port from the environment variables in the pod)

What I was trying to query was the `datahub_usage_event` index which is written to from gms.
`curl 10.100.152.177:8080/datahub_usage_event`
I can see individual datahub usage events indexed but I can't pull out any useful data from them.

**Comment by murdo-moj** [comment]
From my looking at the logs it appears to me that it would be easier to track user behaviour via the frontend logs rather than the backend datahub logs, which are hard to parse. 

It is still possible to get an idea of user behaviour via the datahub logs when usage is low enough that timestamps can act as a link between metadata changes and frontend interactions.

**Comment by murdo-moj** [comment]
https://datahubproject.io/docs/managed-datahub/datahub-api/entity-events-api/ The entity events api has event logs like we'd want, but it's only available in managed datahub
```
{
  "entityUrn": "urn:li:dataset:abc",
  "entityType": "dataset",
  "category": "TAG",
  "operation": "ADD",
  "modifier": "urn:li:tag:PII",
  "parameters": {
    "tagUrn": "urn:li:tag:PII"
  },
  "auditStamp": {
    "actor": "urn:li:corpuser:jdoe",
    "time": 1649953100653   
  }
}
```
